### PR TITLE
TechDraw: Unify Insert View tools (View, ProjGroup, Sheet, Arch, svg, image)

### LINKED ISSUE #5821: [Feature Request] TechDraw: Allow manual entry of the main direction (the front) for a ProjectionGroup
Issue imported from https://tracker.freecad.org/view.php?id=3646

* **Reporter:** vocx
* **Date submitted:** 10/21/2018
* **FreeCAD version:** 0.18
* **Category:** Feature
* **Status:** new
* **Tags:** 

# Original report text

If a body's face is selected, and the user creates a ProjectionGroup, the projection direction is normal to that face. If no face is selected, the projection uses the 3D camera angle.

However, Sketches don't have a face so if the user tries to create a projection normal to that Sketch, he has to rely on adjusting the camera first before calling the ProjectionGroup tool.

An option to manually define the main direction (the front) of the ProjectionGroup could be added, so that selecting a face, or changing the 3D camera is not necessary.

The feature was first requested in the old TechDraw repository https://github.com/WandererFan/FreeCAD-TechDraw/issues/65 by @ceremcem

> [Supposedly introduced with](https://github.com/realthunder/FreeCAD_assembly3/issues/100#issuecomment-421721880) https://github.com/FreeCAD/FreeCAD/commit/aa8feb79d0ab6766f736161f9fdd722ac80817f9, TechDraw generates 3D projection of a sketch by using current camera angle but we can't select a face of a sketch since it has no face.
> ### Proposal
>
> A button (or a context option on the input) to reset the projection angle to `[1.00 0.00 0.00]` would suffice in the task window:
> image

Now it's in the forum https://forum.freecadweb.org/viewtopic.php?f=35&t=31041

# Additional information

OS: Ubuntu 18.04.1 LTS
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.18.14995 (Git)
Build type: Release
Branch: master
Hash: 7b866816785ac1ca9787b63d4bff12873bb3a593
Python version: 2.7.15rc1
Qt version: 4.8.7
Coin version: 4.0.0a
OCC version: 7.3.0
Locale: English/UnitedStates (en_US)

# Other bug information

* **Priority:** normal
* **Severity:** feature
* **Category:** Feature
* **Updated:** 3/7/2021

# Discussion from Mantis ticket

### Comment by @donovaly 2019-01-28 16:16
This issue also popped up in this discussion:
https://forum.freecadweb.org/viewtopic.php?f=35&t=33776&p=282993#p282993

### LINKED ISSUE #12377: [Problem] TechDraw : Insert SVG/picture should not have tools
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Insert SVG/picture should not have tools. They should use the core import feature.

### Full version info

```shell
0.22
```

### Subproject(s) affected?

Techdraw

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #13070: TechDraw: Projection Group use current camera rotation by default
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

One thing that struck me as very weird in techdraw was that proj group uses current camera rotation as the 'front view'. I think it is unintuitive and confusing.

I would suggest that projection group uses the object front view as it's default front-view. If there are usecases to want the camera to be the default front view, then I suggest adding a checkbox 'Use camera as front view' unchecked by default.

Thoughts on that?

### Full version info

```shell
0.22
```

### Subproject(s) affected?

Techdraw

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #13140: TechDraw: 'Insert spreadsheet view' could not need a tool.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

image

In TechDraw we have a tool called 'Insert Spreadsheet view'. This tool could be removed.

The 'Create view' tool should check object type when triggered, and if the selected object is a spreadsheet, then it creates a 'sheet view' object.

I have not tested yet but perhaps the same is true for 'Draft view' and 'arch view'.

@WandererFan thoughts?

### Full version info

```shell
0.22
```

### Subproject(s) affected?

Techdraw

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by PaddleStroke**
1 - We have currently 3 tools that are mutually exclusive. Insert view for shapes, insert spreadsheet and insert arch. The selection of one tool isn't allowed for other tools. Hence it is possible to have one tool inserting all those kind of views.
-> Enable the 'Insert View' tool to create spreadsheet, Arch, SVG and Images views.
-> Remove the Insert spreadsheet and insert arch tools from toolbar.
-> Update the icon to show the insert is more capable.
Note: It is noted that insert spreadsheet could offer a taskbox with more functionality and would thus warrant a taskbox. But the tool currently does not offer such possibility. When the Insert spreadsheet tool is improved and has a proper taskbox letting user type in tables directly for instance, then the tool can be added to the toolbar.

2 - On top of that the insert tool does nothing if nothing is selected besides throwing an error. Instead if nothing is selected, it can open a file browser to handle insertion of image and SVG.
-> Enable the insert View tool to insert SVG and Image when there is no selection.
-> Remove the Insert Image and Insert SVG from toolbar

3 - Inserting a Part view is basically inserting a proj group with only one view. Also Insert Part view does not have a taskbox letting you select the direction. Merging both tool solve both.
-> Insert tool/double clicking a part view opens the Proj Group Task. 
-> Remove proj group tool from toolbar

-> Rework the proj group task layout.
-> Set default to front view of the doc
-> Add a 'front view' and a 'camera' buttons. Camera button set direction to current camera (or current selected face)
image
-> enable user to set custom direction and rotate by custom angle.
image

In total this remove 5 tools from the toolbars without loosing any functionality and while making things more intuitive.

Fixes #13140 
Fixes #12377
Fixes #13070
Fixes #5821

### CODE REVIEW

**Review by maxwxyz — CHANGES_REQUESTED** [review]
You can use this if you like, I've made the dashes more uniform
TechDraw_View

**Comment by maxwxyz** [comment]
Could this be merged with the insert projection view or is this to much?

**Comment by PaddleStroke** [comment]
We could go further because the Insert View tool is not defined when selection is a View. So we could define that so that Insert View trigger 'Insert View details' when a view is selected. But maybe that goes too far. Thoughts?

File: src/Mod/TechDraw/Gui/Command.cpp [hunk 1]
```diff
@@ -287,7 +287,9 @@ CmdTechDrawView::CmdTechDrawView() : Command("TechDraw_View")
     sAppModule = "TechDraw";
     sGroup = QT_TR_NOOP("TechDraw");
     sMenuText = QT_TR_NOOP("Insert View");
-    sToolTipText = QT_TR_NOOP("Insert a View");
+    sToolTipText = QT_TR_NOOP("Insert a View in current page.\n"
```
Comment (maxwxyz) [inline src/Mod/TechDraw/Gui/Command.cpp:290]: Insert a view in current page.
Selected objects, spreadsheets or Arch WB section planes will be added with the current camera orientation.
Without a selection, a file browser lets you select a SVG or image file.

**Comment by maxwxyz** [comment]
> We could go further because the Insert View tool is not defined when selection is a View. So we could define that so that Insert View trigger 'Insert View details' when a view is selected. But maybe that goes too far. Thoughts?

That would be nice, but could it be possible also to do it with the first step? Set projection views and distance, if a 3D object is selected (not spreadsheet or section view from Arch)

**Comment by PaddleStroke** [comment]
@maxwxyz what do you call projection view? Is it projection group? If so I don't think we can because input is the same.
Or is it 'project shape'? In which case I do not understand what this tool is about.

**Comment by maxwxyz** [comment]
@PaddleStroke yes, projection group. Could be called when a shape is selected, default without other views and the result on the page is the same as insert view (but the view is in a projection group)

**Comment by PaddleStroke** [comment]
That makes sense. We can even create a normal view (instead of a proj group) if the user only select the front view.

**Comment by maxwxyz** [comment]
> That makes sense. We can even create a normal view (instead of a proj group) if the user only select the front view.

as long as the  user can create a projection from this normal view later on?

**Comment by PaddleStroke** [comment]
So I have merged projection group as well and it's great.
Now using Insert View on a shape will open the proj-group task panel BUT will still create a normal view until you click on one secondary view. At which point it will replace the normal view by a proj group.

- Normal view now have a taskbox, letting you set scale and direction.
- Double clicking view in the tree now open the task to let you edit.
- Remove one tool from toolbar, simplifying ui.

image

**Comment by maxwxyz** [comment]
@PaddleStroke great! And you also had some suggestion for default or standard views as option for that dialog.

**Comment by PaddleStroke** [comment]
@WandererFan I'm trying to tackle https://github.com/FreeCAD/FreeCAD/issues/5821 at the same time.
It appears to me that the Direction + DirectionX actually defines a placement if I'm not mistaken. So I think it could be good to use the Placement dialog for that. So I turned the middle lineEdit into a button and clicking it opens the placement dialog.
image

What do you think about this? Also is there already built in function to turn Direction/DirectionX into placement?

**Comment by PaddleStroke** [comment]
Edit: Direction/DirectionX seem to rather define a rotation than a placement.

**Comment by maxwxyz** [comment]
@PaddleStroke few things I've noticed, some might not be related to the changes:
1. If possible, switch to the Page view when a body is selected and you click insert view. Then you directly see the results. 
2. I like the behavior of the combined tool but think new users will find it hard to create a view as it is not directly clear that they need a pre-selection and doing so while the sheet is not visible. Maybe a dialog could pop up before showing the file browser if the user wants that or if the user wants to create a view from existing objects?
3. The page currently does not auto expand in the tree view, not sure if this was the same before
4. Maybe a double click on the view in on the page can also open the dialog as with a double click on the tree view
5. When you insert a single view and start creating dimensions and then double click and set a secondary view, all dimensions are gone. Same with section view. Basically all things which are depending of the first view are deleted or get broken when you double click it and create a projection group out of it. Maybe it should always create a projection view?

**Comment by PaddleStroke** [comment]
1. Yes, I usually select in the tree so I have not come accross this yet but it sounds reasonable. This is not something that is modified by my PR though and I'm not sure how to switch to the page. But I can probably find easily. Ah yes it must happen when doubleclicked().
2. Here can you please clarify? Are you talking about the tool -> selection workflow? ie when user trigger the tool before selecting?
3. I don't think I changed that.
4. Sounds reasonable. But I'm not sure know how to do this properly. Let's perhaps keep that for later as to not overload this PR.
5. Ah yes as it deletes the normal view and creates a new one. Always a proj group would definitely solve this and would be much simpler (basically most of my PR goes away), but that kind of bloat the tree if you want a simple view. But maybe it's still better.
6. I do not know how the activeView works. I will look into it.
7. I'm not sure how to handle the draft tool because it takes the same input as normal view. So there's no clear way how to differentiate them.

Btw do you have any idea of icon that could replace the camera icon I put in the button for the 'camera direction' ?

**Comment by maxwxyz** [comment]
> 
>     2. Here can you please clarify? Are you talking about the tool -> selection workflow? ie when user trigger the tool before selecting?

The tool is working fine, but I am not sure that new users get along. Currently, with the separated tools, there is a warning when the user wants to insert a view from a 3D shape but has no selection. In the new combined tool, without a selection just the file dialog opens. I think many new users who want to insert a 3D view do not know that you need to go the 3D view (page is not visible), select your objects, and then click the insert view icon. 
Maybe there could be a warning like today or a hint when no selection is present: You have no selection to insert a view from. Do you want to open a file browser to insert an image file) Yes/No


> 
> Btw do you have any idea of icon that could replace the camera icon I put in the button for the 'camera direction' ?

Currently not, but I think it is clear.

**Comment by WandererFan** [comment]
> Edit: Direction/DirectionX seem to rather define a rotation than a placement.

DirectionX can rotate the view, and it can also reverse (mirror) it.  In general, the user should never need to touch DirectionX.

**Comment by PaddleStroke** [comment]
@WandererFan from what I understand the view can be defined by Direction + an angle (that is modified by the rotate 90 buttons)
I was thinking that the angle is defined by the DirectionX, but I'm not sure this is correct.

**Comment by PaddleStroke** [comment]
@maxwxyz 
1. Done
2. First time you trigger the insert tool without a selection you get a message with a 'do not show again' checkbox
5. I have added a parameter 'InsertAsProjGroup' for testing purposes. If enable the insert tool will create projection group directly. If not enabled it will create simple views. Which can later be turned into proj group at the cost of loosing dimensions / cosmetic etc.

**Comment by PaddleStroke** [comment]
@WandererFan could you please check the following video presenting the point 5. raised by Max. And let me know what you think is best?
https://github.com/FreeCAD/FreeCAD/assets/19984177/6e0cd51f-3242-4497-898a-0e1ab0e7b4ab

- Is there any usecase where the user want a 'DrawViewPart' instead of a projection group with only front?
- What option do you think is best between the tool creating directly a proj group or a DrawViewPart ?

**Comment by PaddleStroke** [comment]
Technically they seems to behave the same. But I understand that it's not as neat having a projection group instead of just a view object.

Perhaps then @maxwxyz we just keep as I made initially, and we add a warning when **editing** a simple view and the user try to add a secondary view, we show a message saying "If you turn your view in a projection group you will loose existing cosmetic and dimensions".

**Comment by maxwxyz** [comment]
That should be in the first attempt already. As then the user needs to chose another angle from the beginning to get the projection group. Otherwise it will be lost.

**Comment by PaddleStroke** [comment]
I found the better solution. The command now creates DrawProjGroupItem which is a subclass of DrawViewPart, and so when the user wants to add secondary projection, we just move the DrawProjGroupItem into the projection group.
Required a bit of changes to DrawProjGroupItem so that it acts as a DrawViewPart when not in a proj group but that's now working well.
I'm even going to implement the other way around, if you unselect secondary views it will go back to a simple DrawProjGroupItem .

File: src/Mod/TechDraw/Gui/Command.cpp [hunk 2]
```diff
@@ -357,64 +408,99 @@ void CmdTechDrawView::activated(int iMsg)
...
-        QMessageBox::warning(Gui::getMainWindow(), QObject::tr("Wrong selection"),
-                             QObject::tr("No Shapes, Groups or Links in this selection"));
-        return;
-    }
+        if (!viewCreated) {
+            // If nothing was selected, then we offer to insert SVG or Images files.
+            bool dontShowAgain = hGrp->GetBool("DontShowInsertFileMessage", false);
+            if (!dontShowAgain) {
+                QMessageBox msgBox;
+                msgBox.setText(msgBox.tr("Note: If you want to insert a shape you need to select it before starting the tool, else it opens a file browser to insert SVG or Images."));
```
Comment (maxwxyz) [inline src/Mod/TechDraw/Gui/Command.cpp:418]: Maybe: 
"If you want to insert a view from existing objects, please select them before evoking this tool. Without a selection, a file browser will open, to insert a SVG or image file."

As it's not only shapes but also spreadsheets, arch, 2D faces,...
also just image, multi selection is not possible in the file browser on my side.

Is it possible to have a "go back" button as well, if a user mistakenly did not have a selection and just wants to abort and select something? Currently there is only OK and then the file browser opens.

**Review by maxwxyz — CHANGES_REQUESTED** [review]
Works very good! I've added one remark on the wording for the dialog.
I've played around a bit and switched between single view and projection group and now, somehow I am not able to move the base view of the projection view anymore. Not sure if related to this PR.
Here is the file:
[test.zip](https://github.com/FreeCAD/FreeCAD/files/14892701/test.zip)

It seems to be reproducible if I create first a standard view without secondary views and then switch it to a projection group. The initial first view becomes unmovable, even via the X/Y values in the property editor (not editable). 

There is also sometimes an issue when switching between projection methods (first/third angle) when I have multiple projections already activated: Sometimes there are 2 checkmarks in the dialog but only one view visible on the sheet. Sometimes when clicking off an on a secondary projection, it appears directly on top of the first view. Also, when switching on/off secondary projections (changing checkmarks in the task dialog), the bottom part vanishes and I cannot set projection method or the spacing: 
image

Otherwise great!

**Comment by PaddleStroke** [comment]
@yorikvanhavre do you mean we just remove draft and Arch views from TD toolbar completely?

**Comment by yorikvanhavre** [comment]
@PaddleStroke in fact yes, they could probably be removed (only the BIM command would exist), but i guess people will complain. I would maybe keep an unified version.

The arch view command is easy to unify with another, because it works only if the selection is an arch sectioneplane or an arch buildingpart.

The draft view is more complicated because it works on groups too and the user needs to decide if they prefer to use the draft view or the normal TD view.

i would say either arch and draft are joined together like in BIM, or arch is unified with the others like you propose, and draft stays standalone. i have no real preference... the first one would have the advantage of being coherent with BIM

**Comment by PaddleStroke** [comment]
Currently I have done exactly as you said: arch view is in the unified tool because it's easy (specific selection type) but the draft view is difficult because it's actually not specific to one kind of selection. You can for example create a PartDesign Body and use it to make a draft view.

So I do not know what to do with this. So for now I have left it as a separate tool. But it's annoying me :)

I was thinking that we could detect draft objects in the selection, and if detected prompt the user : 
"Draft objects detected, do you want to make a draft view or a normal view?"
My issue is that I don't know which objects should be considered and how to detect them. As I have little knowledge about draft it's hard to say what is the usecase of the Draft View, and so to determine which objects should be detected.

Perhaps if you can tell me for what objects we should offer the draft view ?

Otherwise the draft view tool could be moved to Draft workbench. But I don't know if this is a good idea.

**Comment by PaddleStroke** [comment]
@maxwxyz thanks for your comment, sorry I could not fix those yet as I'm on assembly fixing some annoying things.

**Comment by FEA-eng** [comment]
> but i guess people will complain

Yes, I would complain ;-) I use Draft view in TD as a workaround to not being able to manually modify the TD views. I would appreciate a unified tool as proposed in this PR though.

**Comment by yorikvanhavre** [comment]
> Currently I have done exactly as you said: arch view is in the unified tool because it's easy (specific selection type) but the draft view is difficult because it's actually not specific to one kind of selection. You can for example create a PartDesign Body and use it to make a draft view.

Yes I think that's the best solution for now. The problem is as @FEA-eng says sometimes people want to use the draft view even with non-draft content. So it's hard to make it work "automatically", so I have no better idea at the moment..

**Comment by PaddleStroke** [comment]
@FEA-eng what do you mean by 'modifying' by the way?

**Comment by FEA-eng** [comment]
> @FEA-eng what do you mean by 'modifying' by the way?

Mainly using sketching tools to add lines to the view but also removing some lines, replacing them with others and so on. Basically, working on the TD view as if it was a sketch. This is possible in other software like SolidWorks. In FreeCAD, you basically have to use the [Draft Shape2DView](https://wiki.freecad.org/Draft_Shape2DView) tool for that and then create a view of the sketch or draft. An example is presented here: https://forum.freecad.org/viewtopic.php?t=68924

**Comment by FEA-eng** [comment]
> The problem is as @FEA-eng says sometimes people want to use the draft view even with non-draft content.

To make it clear, I'm thinking about the workflow described in the reply to Paddle - using the Draft Shape2DView tool to create an editable replacement for TD view and then creating an actual TD view of this draft or sketch. But Draft and Arch views in TD could be useful in other cases too.

**Comment by PaddleStroke** [comment]
@maxwxyz I have fixed all the problems you mentioned. And updated the text as you offered.

File: src/Mod/TechDraw/Gui/Command.cpp [hunk 3]
```diff
@@ -357,64 +408,99 @@ void CmdTechDrawView::activated(int iMsg)
...
-        QMessageBox::warning(Gui::getMainWindow(), QObject::tr("Wrong selection"),
-                             QObject::tr("No Shapes, Groups or Links in this selection"));
-        return;
-    }
+        if (!viewCreated) {
+            // If nothing was selected, then we offer to insert SVG or Images files.
+            bool dontShowAgain = hGrp->GetBool("DontShowInsertFileMessage", false);
+            if (!dontShowAgain) {
+                QMessageBox msgBox;
+                msgBox.setText(msgBox.tr("If you want to insert a view from existing objects, please select them before evoking this tool. Without a selection, a file browser will open, to insert a SVG or image file."));
```
Comment (chennes) [inline src/Mod/TechDraw/Gui/Command.cpp:418]: "Evoking" made me chuckle... you mean "invoking" here....
```suggestion
                msgBox.setText(msgBox.tr("If you want to insert a view from existing objects, please select them before invoking this tool. Without a selection, a file browser will open, to insert a SVG or image file."));
```

**Review by chennes — APPROVED** [review]
I gave the unified insert tool a try tonight and I like it very much -- nice work, @PaddleStroke.

**Review by maxwxyz — APPROVED** [review]

**Comment by WandererFan** [comment]
Can these commits be squashed, or do we need to keep them as is?

**Comment by PaddleStroke** [comment]
@WandererFan as you prefer. I tried to make commits that make sense, so perhaps it is best to keep them separate.

**Comment by WandererFan** [comment]
This needs some wiki changes.

**Pull request merged**